### ISSUE TITLE
Automate tracking/inclusion of common files and binary files

### ISSUE THREAD

**Issue by ganwell** [fsfe/reuse-tool#72]
TLDR: I read #28 and the example https://git.io/fj577. I am fine with that, I think we need a way to automate this. And for some files the tagging has to be out-of-band, headers wont work.

# Idea

## Singular files

My proposal:

```reuse set-license --license X --copyright Y $GLOB```

Which would edit `.reuse/dep5`, also replace entries if needed and report conflicts.

For example:

```reuse set-license --license CC-BY-SA-4.0 --copyright "2019 Ganwell" static/*.png```

Adds this to .reuse/dep5:

```plain
Files: static/*.png
Copyright: 2019 Ganwell
License: CC-BY-SA-4.0
```

## Common files

```reuse set-common-config --license AGPL-3.0-or-later --copyright "2019 Ganwell"```

This should add entries to .reuse/dep5 for common files like .gitignore, Cargo.tom, package.json. It's just too much to ask from users to track all these files. I also don't like headers in these file, because many of these files are **generated/edited by command-line tools** and these headers have a high chance to break the file, or the **command-line tools might remove the header** again.

# Motivation

Here is my motivation, I kind of argue that not every file should be tracked, but I can accept every file being tagged when there is good support for that.

I understand the goal to include every file. It feels like a good goal to me too. But I am not sure if really everything should be included. I moved 20 repositories to AGPL yesterday and I refused to add a license/copyright indicators to everything, it took over 10h regardless. Two mayor cases:

## Config

1. Config files, these are often strongly structured: Cargo.tom, package.json, .travis-ci.yml, .gitlab-ci.yml. A notable exception  I see: setup.py is actually code and I sometimes write code in these that should be under copyright/copyleft.

2. Less structure config files and config for IDEs and helper tools, like: .gitignore, .flake8, setup.cfg. In my option they don't hold a copyright-able work. Although in theory maybe everything does, but really... its just a burden.

## Binary files

I think reuse should offer to set license/copyright on them, but by automatically extending .reuse/dep5 or some other out-of-band file. I can't add meaningful copyright indicators to files like *.png, *.jpg, *.dat. Some might actually have a comment field inside, but

1. I have to start the programm that produced that file to edit the comment

2. Almost nobody would ever bother to look inside a binary file

**Comment by carmenbianca** [comment]
Hey @ganwell Thanks for the issue.

Have you read this? https://reuse.software/tutorial/#binary-and-uncommentable-files

The way to license files that you can't add comments to, is to create a file "mypicture.png.license", and add the header to that `.license` file. That works just as well for .gitignore etc etc etc. As for adding the copyright and licensing information *into* the binary file, that sounds like a hassle, as you stated.

> `reuse set-license --license CC-BY-SA-4.0 --copyright "2019 Ganwell" static/*.png`

I can appreciate this, but globs are sadly expanded by the shell. This makes it non-trivial to generate glob entries for DEP5 from the command line. Because adding entries to DEP5 is straightforward already, I'm not sure how valuable this would be.

> `reuse set-common-config --license AGPL-3.0-or-later --copyright "2019 Ganwell"`

I can appreciate this as well, and it wouldn't be very difficult to implement this, but like the other one, I'm not certain how laborious it is to manually edit `.reuse/dep5`, the header, or execute `reuse addheader --copyright bla --license bla .gitignore setup.cfg etc etc etc`.

**Comment by ganwell** [comment]
Hi @carmenbianca 

Thanks for your answer.

> The way to license files that you can't add comments to, is to create a file "mypicture.png.license", and add the header to that `.license` file. That works just as well for .gitignore etc etc etc. As for adding the copyright and licensing information _into_ the binary file, that sounds like a hassle, as you stated.

Sorry, that is a bit nit-picky, but I am quite sure "mypicture.png.license" is never going to work. Software-Developers are artists, I don't know a single developer who would do this to his repository. This is so unaesthetic, no one would do that. People go out of their way to hide all distracting / ugly files in hidden directories. I can barely accept the headers in source files. It has to be in the DEP5 file and I think that is good. (Next paragraph) Thank you for pointing that out. 

> I can appreciate this, but globs are sadly expanded by the shell. This makes it non-trivial to generate glob entries for DEP5 from the command line. Because adding entries to DEP5 is straightforward already, I'm not sure how valuable this would be.

Yes you are right. Is reuse currently checking the DEP5 files for conflicts (conflicting globs [1]) and dead entries? My main concern is that I end up with conflicting and outdated entries.

> I can appreciate this as well, and it wouldn't be very difficult to implement this, but like the other one, I'm not certain how laborious it is to manually edit `.reuse/dep5`, the header, or execute `reuse addheader --copyright bla --license bla .gitignore setup.cfg etc etc etc`.

From my experience moving these 20 Repositories, yesterday (actually I would have moved more, if it was easier). There is overwhelming amount of meta-data files. Some wizard like mode proposing the right thing to do for commonly known files, would be nice. Maybe during `reuse init`.

[1] One could probably just expand the globs and see if a file is in multiple globs or a glob returns nothing.

**Comment by ganwell** [comment]
Update, I complained about a feature that is optional, there is no need to do that: text removed.

**Comment by ganwell** [comment]
I realized that what that looks like a conflict, could also be intentional. I guess think the best thing to do, is:

* a carefully crafted dep5 template for config/meta-data files

* addheader for actual source and content

* manual additional entries in dep5 for binary files

I close this in favor of an issue about consistency of lint results.
